cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp

```diff
@@ -157,6 +161,53 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
             ;
     }
 
+    static void copy_array_using_ellipsis(wrapped_type & self, pybind11::array & arr_in)
+    {
+        const auto left_shape = self.shape();
+        const auto * right_shape = arr_in.shape();
+
+        auto throw_error = [=]()
+        {
+            std::ostringstream msg;
+            msg << "Input array from shape(";
+            for (pybind11::ssize_t i = 0; i < arr_in.ndim(); ++i)
+            {
+                msg << right_shape[i];
+                if (i != arr_in.ndim() - 1)
+                {
+                    msg << ", ";
+                }
+            }
+            msg << ") into shape(";
+            for (size_t i = 0; i < self.ndim(); ++i)
+            {
+                msg << left_shape[i];
+                if (i != self.ndim() - 1)
+                {
+                    msg << ", ";
+                }
+            }
+            msg << ")";
+
+            throw std::runtime_error(msg.str());
+        };
+
+        if (self.ndim() != static_cast<size_t>(arr_in.ndim()))
+        {
+            throw_error();
+        }
+
+        for (size_t i = 0; i < left_shape.size(); ++i)
+        {
+            if (left_shape[i] != static_cast<size_t>(right_shape[i]))
+            {
+                throw_error();
+            }
+        }
+
+        std::memcpy(self.data(), arr_in.data(), arr_in.nbytes());
```

Comment:
I found that we also need to support broadcasting array with different types, which is implemented in Numpy:

```py
ndarr = np.arange(2*3*4, dtype='float32').reshape((2, 3, 4))
ndarr2 = np.arange(2*3*4, dtype='int32').reshape((2, 3, 4))
ndarr[...] = ndarr2[...]
```

I try to implement this:
```cpp
if  (dtype_is_type<T>(arr_in))
{
  std::memcpy(self.data(), arr_in.data(), arr_in.nbytes());
}
else
{
  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
  const auto * ptr = reinterpret_cast<const int*>(arr_in.data()); // should decide the type from arr_in

  for (size_t size = 0; size < self.size(); ++size)
  {
      T * it = self.begin() + size;
      *it = *(ptr + size);
  }
}
```
Now I just use `reinterpret_cast<const int*>` for demo, but I found that I cannot get the data type directly from `arr_in`.
I can get the type name by `pybind11::array::dtype()`, so it's possible to write the code like the following which is ugly:
```cpp
if(arr_in.dtype() == "int32") {
  const auto * ptr = reinterpret_cast<const int*>(arr_in.data()); 

  for (size_t size = 0; size < self.size(); ++size)
  {
      T * it = self.begin() + size;
      *it = *(ptr + size);
  }  
} else if (arr_in.dtype() == "float32") {
  const auto * ptr = reinterpret_cast<const float*>(arr_in.data());
...

Is there a better way to achieve it?

Reply:
I would be surprised if `reinterpret_cast<float *>(pointer_to_int)` work for the type casting.  It should need a `static_cast`.  I am not sure how numpy implements the broadcasting.  But perhaps you can reference its approach?

I would guess broadcasting support is significantly more complex than ellipsis.  It might be a good idea to use a separate PR for it.

Reply:
Sorry I misinterpret your intention of using `reinterpret_cast`.  It's for interpreting the input ndarray and it's fine.

Reply:
In addition to learn from numpy, perhaps you can also try to build a mapping table for the type broadcasting.  A class or class template might probably help implementation:

```cpp
template <typename D /* for destination type */> struct TypeBroadCast
{

// Populating the map should be straight-forward.
std::map<std::string /* dtype string */, std::type_info> m_type_map;

D * broadcast(pybind11::array const & arr)
{

// I don't know how to implement the conversion from dtype to static type D.

}

};
```

Reply:
The problem is when the `__setitem__` is called, the parameter is `pybind11::array` ([ref](https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L680)) which doesn't return any type information so that I cannot use decltype like `reinterpret_cast<decltype(arr_in.sth)>`,
So now maybe the ugly implementation (as the example above) could be a workaround and we can improve it in the future?

Reply:
We can improve it in the future, but do not need a workaround at the moment.  The type broadcasting is not part of the original request.  We can (and maybe should) still do it, but it may be done using a separate PR.

pybind11 should also contain code for the same purpose.  Perhaps you can take a look at https://github.com/pybind/pybind11/blob/479e9a50f38cfa8e525a697d7f5d850f851223b5/include/pybind11/numpy.h#L1016

Reply:
`pybind11:array_t<T>` do have the type information, however `__setitem__` only provide `pybind11:array` at the moment. `array_t` is inherited from `array`

it's possible to do this but have the same issue
```
            .def(
                "__setitem__",
                [](wrapped_type & self, const py::ellipsis &, pybind11::array_t<int> & arr_in)
                ...
            .def(
                "__setitem__",
                [](wrapped_type & self, const py::ellipsis &, pybind11::array_t<float> & arr_in)
                ...
  ```

Reply:
What is the reason to use `std::memcpy` but not `std::copy_n`?

Reply:
The main reason is that `pybind11:array` doesn't provide iterators, resulting in not supporting `copy_n`.
In this case, since the buffer type and size will be the same, I think `memcpy` should work well.
Even if `std::copy_n` is available, I wonder about the performance difference between calling `memcpy` and `copy_n`?

Reply:
`copy_n` can be used for the buffer just like `memset`.  When using `memset`, I guess you are relying more on how OS chooses to implement it because compiler needs a good reason not to use the OS implementation.  With `copy_n` it's more on the stdlib.

`memset` and `copy_n` have different definition and perhaps the former has a better chance to be optimized away: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=8537 .

I prefer `copy_n`, because it induces less change of code when we move from pointer to an alternate iterator.  Using `memset` at this point requires one to recode it to `copy_n` later.

Reply:
Please note that while it is dangerous to directly use the memory buffer of the input `ndarray`, because the input may use non-simple strides to skip elements, e.g., `SimpleArrayFloat64[...] = brr[::2,:,:]`.  It would be interesting to add a unit test for the strided input.

If the test fails, I am still fine to use the memory buffer for this feature, as a draft implementation.  But the failed test should be added as a known failure and a new issue should be created to track the fix.

Reply:
resolve type broadcast